File: code/SungHo-Ck/week6/prog_118669.cpp
#include <string>
#include <vector>
#include <algorithm>
#include <queue>
#include <iostream>
#define INF 987654321

using namespace std;

vector<int> solution(int n, vector<vector<int>> paths, vector<int>gates, vector<int>summits) {
    vector<int> answer = { -1,INF };

    vector<vector<pair<int, int>>> graph(n + 1);
    for (int i = 0; i < paths.size(); i++) {
		graph[paths[i][0]].push_back({ paths[i][1],paths[i][2] });
		graph[paths[i][1]].push_back({ paths[i][0],paths[i][2] });
	}

    vector<int> isSummit(n + 1, 0);
    for (int i = 0; i < summits.size(); i++) {
        isSummit[summits[i]] = 1;
    }

    vector<int> dist(n + 1, INF);
	priority_queue<pair<int, int>> pq;
    for (int i = 0; i < gates.size(); i++) {
		dist[gates[i]] = 0;
		pq.push({ 0,gates[i] });
	}
    while (!pq.empty()) {
		int cur_dist = pq.top().first;
        int cur = pq.top().second;
		pq.pop();
        if (dist[cur] < cur_dist || isSummit[cur]) {
            continue;
        }
        for (int i = 0; i < graph[cur].size(); i++) {
			int next = graph[cur][i].first;
            int next_dist = max(dist[cur], graph[cur][i].second);
            if (dist[next] > next_dist) {
				dist[next] = next_dist;
				pq.push({ next_dist,next });
			}
		}
	}
    sort(summits.begin(), summits.end());
    for (int i = 0; i < summits.size(); i++) {
        if (answer[1] > dist[summits[i]]) {
			answer[0] = summits[i];
			answer[1] = dist[summits[i]];
		}
	}
	return answer;
}

int main(void) {
    vector<vector<int>> paths = { {1, 2, 3} ,{2, 3, 5},{2, 4, 2},{2, 5, 4},{3, 4, 4},{4, 5, 3},{4, 6, 1},{5, 6, 1} };
    vector<int> gates = { 1,3 };
    vector<int> summits = { 5 };
    vector<int> ans = solution(6, paths, gates, summits);
    cout << ans[0] << "," << ans[1] << endl;

    paths = { {1, 4, 4} ,{1, 6, 1},{1, 7, 3},{2, 5, 2},{3, 7, 4},{5, 6, 6} };
    gates = { 1 };
    summits = { 2,3,4 };
    ans = solution(7, paths, gates, summits);
    cout << ans[0] << "," << ans[1] << endl;

    paths = { {1, 2, 5} ,{1, 4, 1},{2, 3, 1},{2, 6, 7},{4, 5, 1},{5, 6, 1},{6, 7, 1} };
    gates = { 3,7 };
    summits = { 1,5 };
    ans = solution(7, paths, gates, summits);
    cout << ans[0] << "," << ans[1] << endl;

    paths = { {1, 3, 10} ,{1, 4, 20},{2, 3, 4},{2, 4, 6},{3, 5, 20},{4, 5, 6} };
    gates = { 1,2 };
    summits = { 5 };
    ans = solution(5, paths, gates, summits);
    cout << ans[0] << "," << ans[1] << endl;
    return 0;
}

File: code/jonghyeonjo99/week6/prog_118669.py
import heapq
from math import inf

def solution(n, paths, gates, summits):
    # 간선 정리 (양방향)
    graph = [[] for _ in range(n + 1)]
    for i, j, w in paths:
        graph[i].append([j, w])
        graph[j].append([i, w])

    # 산봉우리 판별
    is_summit = [False] * (n + 1)
    for summit in summits:
        is_summit[summit] = True

    # gates 모두 시작 위치
    distance = [inf] * (n + 1)
    queue = []
    for gate in gates:
        distance[gate] = 0
        heapq.heappush(queue, [0, gate])

    # 다익스트라
    while queue:
        d, i = heapq.heappop(queue)
        # 산봉우리면 바로 continue
        # 이렇게 해야 두 개 이상의 산봉우리를 방문하지 않는다.
        if distance[i] < d or is_summit[i]:
            continue
        for j, dd in graph[i]:
            dd = max(distance[i], dd)
            if distance[j] > dd:
                distance[j] = dd
                heapq.heappush(queue, [dd, j])

    # 결과
    # 거리가 같으면 산봉우리의 번호가 작은 것을 출력해야 하므로
    # 산봉우리를 정렬하여 살펴보자.
    result = [-1, inf]
    for summit in sorted(summits):
        if distance[summit] < result[1]:
            result[0] = summit
            result[1] = distance[summit]
    return result